addons/hr_org_chart/static/src/views/hierarchy/hierarchy_renderer.js

```diff
@@ -0,0 +1,123 @@
+                    }
+                },
+                onDrop: (params) => {
+                    this.nodeDrop(params);
+                },
+                onElementEnter: ({ addClass, element }) => {
+                    addClass(element, "o_hierarchy_hover");
+                },
+                onElementLeave: ({ removeClass, element }) => {
+                    removeClass(element, "o_hierarchy_hover");
+                },
+                onRowEnter: ({ addClass, row }) => {
+                    addClass(row, "o_hierarchy_hover");
+                },
+                onRowLeave: ({ removeClass, row }) => {
+                    removeClass(row, "o_hierarchy_hover");
+                },
+            });
+        }
+        onPatched(this.onPatched);
+    }
+
+    onPatched() {
+        const row = this.rendererRef.el.querySelector(":scope .o_hierarchy_row:last-child");
+        if (row) {
+            const scrollable = this.env.isSmall ? document.querySelector(".o_hierarchy_view") : this.rendererRef.el.parentNode;
+            scrollable.classList.add(".o_hierarchy_auto-scroll");
+            scrollTo(row, { scrollable, isAnchor: true });
+            scrollable.classList.remove(".o_hierarchy_auto-scroll");
+        }
+    }
+
+    get canDragAndDropRecord() {
+        return this.draggable && !this.env.isSmall;
+    }
+
+    get draggable() {
+        return this.props.archInfo.draggable;
+    }
+
+    get rows() {
+        const rootNodes = this.props.model.root.rootNodes;
+        const rows = [{ nodes: rootNodes }];
+        const processNode = (node) => {
+            if (!node.isLeaf) {
+                rows.push({ parentNode: node, nodes: node.nodes});
+                for (const subNode of node.nodes) {
+                    processNode(subNode);
+                }
+            }
+        };
+
+        for (const node of this.props.model.root.rootNodes) {
+            processNode(node);
+        }
+
+        return rows;
+    }
+
+    async nodeDrop({ element, row, nextRow, newParentNode }) {
+        let parentNodeId, parentResId;
+        if (newParentNode) {
+            parentNodeId = newParentNode.dataset.nodeId;
+        } else if (nextRow?.dataset.rowId !== row.dataset.rowId) {
+            parentNodeId = nextRow.dataset.parentNodeId;
+            if (!parentNodeId) {
+                const nodes = this.rows[nextRow.dataset.rowId].nodes || [];
+                if (nodes) {
+                    parentNodeId = nodes[0].parentNode?.id;
+                    if (!parentNodeId) {
+                        parentResId = nodes[0].parentResId;
+                        if (!nodes.every((node) => node.parentResId === parentResId)) {
+                            // TODO: display notification to alert the user we don't find a parent to correctly move the record
+                            return;
+                        }
+                    }
+                }
+            }
```

Comment:
It maybe possible to reduce the number of if ? It s complexe to read it

Reply:
I will check if I can reduce that.